### PR TITLE
Irc patch

### PR DESCRIPTION
**Pull request by AlexHeide**
- [x] Fix printing issue for IRCs in CustomHelper
- [x] Standardize serialization of classes to dict for restarting
- [x] Fix serialization of IRCs (iterations not tracked correctly)
- [x] Add CustomHelper tests for IRCS and Linesearch
- [x] Requesting `irc_points = 1` now quits immediately after converging 1 point
   The previous behavior would add the TS as the zeroth point and quit. The immediate quit means that the wfn and energy
   match the final point when run in the psi4 driver as well.
- [x] Additional IRC convergence check. To prevent stepping over the minimum a negative overlap and energy increase
   terminates the IRC. This is in addition to an overlap of less than -0.7. Fixes 1 IRC test in psi4 where two points with higher energy
   beyond the minimum are optimized.

### CODE REVIEW

File: optking/opt_helper.py [hunk 1]
```diff
@@ -567,13 +571,13 @@ def __init__(self, optimization_input, **kwargs):
...
         self.computer_type = "qc"
         self.build_coordinates()
-        print(type(self.params))
         self.opt_manager = OptimizationManager(self.molsys, self.history, self.params, self.computer)
 
     @classmethod
     def from_dict(cls, d):
         helper = super().from_dict(d)
         helper.computer = compute_wrappers.make_computer_from_dict("qc", d.get("computer"))
+        helper.opt_manager = OptimizationManager.from_dict(d["opt_manager"], helper.molsys, helper.history, helper.params, helper.computer)
```
Comment (psi-rking) [inline optking/opt_helper.py:580]: Please check the comment block at the top of opt_helper.py and either update it, or remove it to the extent that it's inaccurate.

File: optking/optimize.py [hunk 2]
```diff
@@ -107,18 +107,66 @@ def __init__(self, molsys: Molsys, history_object: history.History, params: op.O
...
+    @classmethod
+    def from_dict(cls, d, molsys, history, params, computer):
+        """ Reload attributes from the provided dictionary. Create all necessary classes.
+        
+        To prevent duplication, OptHelper handles converting the molsys, history, params, and computer to/from dict
+        """
+
+        manager = cls(molsys, history, params, computer)
+        manager.direction = d["direction"]
+        method = "IRC" if params.opt_type == "IRC" else params.step_type
```
Comment (psi-rking) [inline optking/optimize.py:151]: In the future, we may want to support method == "IRC" and also step_type = "NR"/"RFO",  etc..  Currently unsupported?

**Review by psi-rking — DISMISSED** [review]

File: optking/IRCdata.py [hunk 3]
```diff
@@ -56,22 +56,28 @@ def add_pivot(self, q_p, x_p):
...
-            "Energy": self.energy,
-            "Pivot Intco Values": self.q_pivot.tolist(),
-            "Pivot Geometry": self.x_pivot.tolist(),
-            "Step Distance": self.step_dist,
-            "Arc Distance": self.arc_dist,
-            "Line Distance": self.line_dist,
+            "step_number": self.step_number,
+            "q": self.q.tolist(),
+            "x": self.x.tolist(),
+            "f_q": self.f_q.tolist(),
```
[the remark this answers is not in the document]
Comment (AlexHeide) [reply optking/IRCdata.py:64]: That is a good point I think I have referenced that poorly in the documentation in a few spots.

File: optking/optimize.py (hunk 2 above)
Comment (AlexHeide) [reply optking/optimize.py:151]: Is this to run an optimization directly following an IRC? I'm unclear what the behavior of an IRC + RFO would be.

File: optking/IRCdata.py [hunk 4]
```diff
@@ -191,7 +219,7 @@ def step_dist(self, step=None):
         index = -1 if step is None else step
         return self.irc_points[index].step_dist
 
-    def test_for_irc_minimum(self, f_q):
+    def test_for_irc_minimum(self, f_q, energy):
         """ Given current forces, checks if we are at/near a minimum
         For now, checks if forces are opposite those are previous pivot point
```
[the remark this answers is not in the document]
Comment (AlexHeide) [reply optking/IRCdata.py]: Updated. And I added a section to the convergence documentation page.

File: optking/optimize.py (hunk 2 above)
Comment (psi-rking) [reply optking/optimize.py:151]: Each point along the IRC is the result of a constrained optimization.  But using the projected forces one could search for the point using line search, NR, or RFO methods, no?  The RFO might be a little weird.  I haven't thought about the IRC algorithm for awhile.

**Review by psi-rking — APPROVED** [review]

File: optking/optimize.py (hunk 2 above)
Comment (psi-rking) [reply optking/optimize.py:151]: I think as implemented, we're using NR.  Implicit in that Eqn. 24 in dq_irc, but I wouldn't swear to it.

Comment (psi-rking) [reply optking/optimize.py:151]: Anyway, not a front-burner issue, and we can merge at will.

**Pull request merged**